src/Compilers/Core/Portable/Operations/Operation.cs

```diff
@@ -46,41 +76,203 @@ public Operation(OperationKind kind, SyntaxNode syntax, ITypeSymbol type, Option
+        /// 
+        /// Use this to create IOperation when we don't have proper specific IOperation yet for given language construct
+        /// </summary>
+        public static IOperation CreateOperationNone(SemanticModel semanticModel, SyntaxNode node, Optional<object> constantValue, Func<ImmutableArray<IOperation>> getChildren)
         {
-            private readonly Func<ImmutableArray<IOperation>> _getChildren;
+            return new NoneOperation(semanticModel, node, constantValue, getChildren);
+        }
+
+        public static T SetParentOperation<T>(T operation, IOperation parent) where T : IOperation
+        {
+            // explicit cast is not allowed, so using "as" instead
+            (operation as Operation)?.SetParentOperation(parent);
+            return operation;
+        }
 
-            public NoneOperation(SyntaxNode node, Optional<object> constantValue, Func<ImmutableArray<IOperation>> getChildren)
+        public static ImmutableArray<T> SetParentOperation<T>(ImmutableArray<T> operations, IOperation parent) where T : IOperation
+        {
+            // check quick bail out case first
+            if (operations.Length == 0)
             {
-                Syntax = node;
-                ConstantValue = constantValue;
-                _getChildren = getChildren;
+                // no element
+                return operations;
+            }
+
+            // race is okay. penalty is going through a loop one more time
+            // explicit cast is not allowed, so using "as" instead
+            // invalid expression can have null element in the array
+            if ((operations[0] as Operation)?._parentDoNotAccessDirectly != null)
+            {
+                // already initialized
+                return operations;
+            }
+
+            foreach (var operation in operations)
+            {
+                // go through slowest path
+                SetParentOperation(operation, parent);
             }
 
-            public OperationKind Kind => OperationKind.None;
+            return operations;
+        }
 
-            public SyntaxNode Syntax { get; }
+        public static T ResetParentOperation<T>(T operation) where T : IOperation
+        {
+            if (operation == null)
+            {
+                return operation;
+            }
 
-            public ITypeSymbol Type => null;
+            Interlocked.Exchange(ref (operation as Operation)._parentDoNotAccessDirectly, null);
+            return operation;
+        }
 
-            public Optional<object> ConstantValue { get; }
+        private class NoneOperation : Operation
+        {
+            private readonly Func<ImmutableArray<IOperation>> _getChildren;
 
-            public void Accept(OperationVisitor visitor)
+            public NoneOperation(SemanticModel semanticMode, SyntaxNode node, Optional<object> constantValue, Func<ImmutableArray<IOperation>> getChildren) :
```

Comment:
typo: s/semanticMode/semanticModel/
You may want to do a broader check on this (I didn't look everywhere).